src/core/datafile/UpgradeTo13MixerRename.cpp

```diff
@@ -0,0 +1,52 @@
+/*
+ * UpgradeMixerRename.cpp
+ *   Functor for upgrading data files after the fxmixer tag rename
+ *   and fxch attribute rename.
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "datafile/UpgradeTo13MixerRename.h"
+
+namespace lmms
+{
+
+
+void UpgradeTo13MixerRename::upgrade()
+{
+	printf("Running Upgrade To 1.3: Mixer Rename\n");
```

Comment:
I think it would be really helpful to be able to enable a "debug/dev mode" in the `DataFile` class. When I did some bug hunting for an issue I added some functionality that printed the resulting XML after every upgrade step so that I could check how it evolves and if something is missed during an update. This debug mode could then also output the step description as above.

Returning a step description could be implemented as a virtual function in the underlying new base class which is then called before each upgrade steps. Printing the XML after each transformation (and perhaps the starting XML) could be done by `DataFile`  itself if it is constructed with being in debug mode.

Reply:
Heh, woops, that was not supposed to make it into the commit, was just a sanity test.

Reply:
> So you are thinking something like virtual DataFileUpgrader::preUpgrade and DataFileUpgrader::postUpgrade ?

That seems a bit much, no? Maybe just make the upgrades store a name and let `DataFile::upgrade` do the printing?